File: Java_ChatGPT/State/Round1/Code.java
package State.Round1;

// State interface defines the methods for actions.
interface State {
    void turnOn(Light light);
    void turnOff(Light light);
}

// Concrete State: OnState
class OnState implements State {
    @Override
    public void turnOn(Light light) {
        System.out.println("The light is already ON.");
    }

    @Override
    public void turnOff(Light light) {
        System.out.println("Turning the light OFF.");
        light.setState(new OffState());
    }
}

// Concrete State: OffState
class OffState implements State {
    @Override
    public void turnOn(Light light) {
        System.out.println("Turning the light ON.");
        light.setState(new OnState());
    }

    @Override
    public void turnOff(Light light) {
        System.out.println("The light is already OFF.");
    }
}

// Light class that manages state transitions
class Light {
    private State state;

    public Light() {
        this.state = new OffState();  // Default state is OFF.
    }

    public void setState(State state) {
        this.state = state;
    }

    public void turnOn() {
        state.turnOn(this);
    }

    public void turnOff() {
        state.turnOff(this);
    }

    public State getState() {
        return state;
    }
}

File: Gemini_Pro/StatePattern/Round1/Code.py
class State:
    """Interface for light states."""

    def turn_on(self, light):
        raise NotImplementedError

    def turn_off(self, light):
        raise NotImplementedError


class OnState(State):
    """Represents the 'On' state of the light."""

    def turn_on(self, light):
        print("Light is already on.")

    def turn_off(self, light):
        print("Turning light off.")
        light.set_state(OffState())


class OffState(State):
    """Represents the 'Off' state of the light."""

    def turn_on(self, light):
        print("Turning light on.")
        light.set_state(OnState())

    def turn_off(self, light):
        print("Light is already off.")


class Light:
    """Represents a light with state management."""

    def __init__(self):
        self._state = OffState()

    def set_state(self, state):
        self._state = state

    def turn_on(self):
        self._state.turn_on(self)

    def turn_off(self):
        self._state.turn_off(self)